### PR TITLE
add in layer gpt2 tokenizer

### PR DESCRIPTION
**Pull request by piEsposito**
# What does this PR do?

- Adds in layer `TFGPT2Tokenizer` to enable serialization and serving it with TF Serving
- Small fixes on Tensorflow generation utils and GPT2 attentions to use `tf.shape` instead of `Tensor.shape` to solve max sequence length; and
- Explicitly unstacking the past key values on TF GPT2 Attention layer to avoid `None` shape issues that come with `generate` on TF compiled models.

Addresses first step of #19992 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. -> https://github.com/huggingface/transformers/issues/19992
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by piEsposito** [comment]
Just figured we have to solve some max lengths on the xla_tests. Changing the way we get the shapes broke some stuff.

**Comment by piEsposito** [comment]
Actually, if you want I can handle the generate stuff on another PR, but IMHO it makes sense to keep it here, as it ensures the model will actually work along with the tokenizer.

**Comment by Rocketknight1** [comment]
Yeah, I would definitely keep any generate fixes that you need in this PR. Overall this looks extremely good, though, and great job figuring out a mapping from the existing BPE tokenizers to TF BPE that works!

**Comment by Rocketknight1** [comment]
Because this is touching the core library (by adding a `keras-nlp` dependency and an `is_keras_nlp_available()` check) and TF `generate()` code, I'm going to ping @sgugger and @gante to take a look too just to make sure it's all okay!

**Comment by piEsposito** [comment]
Yeah @Rocketknight1 actually I'm think that the generate thing can get bigger than this PR. What do you think of me taking it out and handling on another PR?

**Comment by piEsposito** [comment]
Actually, just did it. Let's keep it simple and handle TF hell on another time.

**Comment by Rocketknight1** [comment]
Totally fine with that too - as long as it works for the core model we can add extra features over time. I'd love to see how many models in our codebase we could just copy-paste this new approach to as well!

**Comment by piEsposito** [comment]
@Rocketknight1 it works for the core model as per a test on this PR. Do you have any idea on why the add model like runner test is not passing? It seems related to some strange code quality check on a file I didn't change.

**Comment by Rocketknight1** [comment]
@piEsposito This happens sometimes - it's usually because you just happened to fork the repo at a bad time when that bug was present. The easiest way to fix these issues is just to rebase your PR and force push.

Also, since we're adding a new dependency here I want to wait to get a couple of reviews from the core team, but half the company is missing for Thanksgiving, so things might be a little slow with this PR until Monday! I definitely want to push to get it in for the next release, though.

**Comment by piEsposito** [comment]
@Rocketknight1 happy thanksgiving! 

All right, let me rebase it and we wait for the other folks to review it. Thank you for your support :D.

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -3827,6 +3842,16 @@
     else:
         from .models.bert import TFBertTokenizer
 
+    try:
+        if not is_keras_nlp_available():
+            raise OptionalDependencyNotAvailable()
+    except OptionalDependencyNotAvailable:
+        print("except")
```
Comment (gante) [inline src/transformers/__init__.py:3849]:
```suggestion
```

File: src/transformers/__init__.py [hunk 2]
```diff
@@ -3827,6 +3842,16 @@
...
+        from .utils.dummy_keras_nlp_objects import *
+    else:
+        print("import")
```
Comment (gante) [inline src/transformers/__init__.py:3852]:
```suggestion
```

File: src/transformers/utils/dummy_tf_objects.py [hunk 3]
```diff
@@ -1350,6 +1350,13 @@ def __init__(self, *args, **kwargs):
         requires_backends(self, ["tf"])
 
 
+class TFGPT2Tokenizer(metaclass=DummyObject):
+    _backends = ["tf"]
+
+    def __init__(self, *args, **kwargs):
+        requires_backends(self, ["tf"])
```
Comment (gante) [inline src/transformers/utils/dummy_tf_objects.py:1357]: I don't think this one should exist 🤔 (the dummy import in the main `__init__.py` points to `utils/dummy_keras_nlp_objects.py`)

File: src/transformers/models/gpt2/tokenization_gpt2_tf.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        tf_tokenizer = TFGPT2Tokenizer.from_pretrained("gpt2")
+        ```
+        """
+        tokenizer = GPT2Tokenizer.from_pretrained(pretrained_model_name_or_path, *init_inputs, **kwargs)
+        return cls.from_tokenizer(tokenizer)
+
+    def call(self, x):
+        input_ids = self.tf_tokenizer(x)
+        attention_mask = tf.ones_like(input_ids)
+        return {"attention_mask": attention_mask, "input_ids": input_ids}
```
Comment (gante) [inline src/transformers/models/gpt2/tokenization_gpt2_tf.py:72]: We should also implement `get_config` and `from_config` methods (see https://keras.io/guides/serialization_and_saving/#savedmodel-format)

File: src/transformers/models/gpt2/tokenization_gpt2_tf.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    straight from `tf.string` inputs to outputs.
+
+    Args:
+        vocab (Dict[str, int]): Vocabulary dict for Byte Pair Tokenizer
+        merges (List[str]): Merges list for Byte Pair Tokenizer
+    """
+
+    def __init__(self, vocab: Dict[str, int], merges: List[str]):
+        super().__init__()
+        self.tf_tokenizer = BytePairTokenizer(vocab, merges)
```
Comment (gante) [inline src/transformers/models/gpt2/tokenization_gpt2_tf.py:29]: We may need to set the `sequence_length` argument based on our padding options, no?

**Review by gante — COMMENTED** [review]
In general, looks good! Added a few comments to fix/discuss.

P.S.: the failing tests may be solved if you rebase with master, as @Rocketknight1 mentioned 🤔

File: src/transformers/__init__.py (hunk 1 above)
Comment (piEsposito) [reply src/transformers/__init__.py:3849]: Yeah sorry for that debugging goes brrr sometimes.

Comment (piEsposito) [reply src/transformers/__init__.py]: Done.

File: src/transformers/__init__.py (hunk 2 above)
Comment (piEsposito) [reply src/transformers/__init__.py]: Done.

File: src/transformers/utils/dummy_tf_objects.py (hunk 3 above)
Comment (piEsposito) [reply src/transformers/utils/dummy_tf_objects.py]: Yeah you're right. Done.

File: src/transformers/models/gpt2/tokenization_gpt2_tf.py (hunk 4 above)
Comment (piEsposito) [reply src/transformers/models/gpt2/tokenization_gpt2_tf.py:72]: Should I do that to comply with HF Transformers or Keras? I didn't implement `from_config` because I was following the [TFBertTokenizer](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/tokenization_bert_tf.py) implementation, the `get_config` I confess I forgot.

Should I do `from_config` too or just keep it like TFBertTokenizer?

Comment (piEsposito) [reply src/transformers/models/gpt2/tokenization_gpt2_tf.py:72]: I'm doing both then, I was just making sure. Thank you for your patience and kindness on this code review. That is a much appreciated talent of yours.

**Comment by piEsposito** [comment]
Should finish to address your review early next week. Stable Diffusion v2 got me into the rabbit hole haha.

File: tests/models/gpt2/test_tokenization_gpt2_tf.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            tokenized = self.tokenizer(text)
+            input_ids_dense = tokenized["input_ids"].to_tensor()
+
+            input_mask = tf.cast(input_ids_dense > 0, tf.int32)
+            # input_mask = tf.reshape(input_mask, [-1, MAX_SEQ_LEN])
+
+            outputs = self.model(
+                input_ids=input_ids_dense,
+                attention_mask=input_mask,
+            )["logits"]
```
Comment (sgugger) [inline tests/models/gpt2/test_tokenization_gpt2_tf.py:41]: Should fit in one line.

**Review by sgugger — APPROVED** [review]
Thanks for adding this! LGTM barring @gante 's comments.

**Comment by piEsposito** [comment]
@Rocketknight1 this should be finished today if I'm lucky. Thank you for understanding the red test thing.

File: tests/models/gpt2/test_tokenization_gpt2_tf.py (hunk 6 above)
Comment (piEsposito) [reply tests/models/gpt2/test_tokenization_gpt2_tf.py]: Done.

**Comment by piEsposito** [comment]
@Rocketknight1 by mistake I removed you from the reviews, but I was actually trying to ask you to do one.

@gante i've addressed your review.

File: src/transformers/models/gpt2/tokenization_gpt2_tf.py (hunk 4 above)
Comment (piEsposito) [reply src/transformers/models/gpt2/tokenization_gpt2_tf.py:104]: Done.

File: src/transformers/models/gpt2/tokenization_gpt2_tf.py (hunk 5 above)
Comment (piEsposito) [reply src/transformers/models/gpt2/tokenization_gpt2_tf.py]: Done.

**Review by Rocketknight1 — APPROVED** [review]
Looks perfect to me now, and good job addressing the changes! @piEsposito Do you want me to merge now, or is there anything else you want to do?

**Comment by piEsposito** [comment]
@Rocketknight1 you can merge it now to keep it on the next release.

I will start trying to replicate this to other BPE tokenizers in the sequence.

Thank you, @gante and @sgugger for the kindness and support.
